drivers/watchdog/wdt_nrfx_wdt.c

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <nrfx_wdt.h>
+#include <nrf_clock.h>
+#include <clock_control.h>
+#include <watchdog.h>
+
+#define SYS_LOG_DOMAIN "WDT NRF"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_WDT_LEVEL
+#include <logging/sys_log.h>
+
+#define NRF_WDT_CHANNELS_NUMBER 8
+
+static struct device DEVICE_NAME_GET(wdt_nrf);
+
+/* There are eight reload registers (channels) in Nordic's WDT. */
+static nrfx_wdt_channel_id m_ch[NRF_WDT_CHANNELS_NUMBER];
+
+/* Each activated and not reloaded channel can generate watchdog interrupt.
+ * Array m_callbacks provides storing callbacks for each channel.
+ */
+static wdt_callback_t      m_callbacks[NRF_WDT_CHANNELS_NUMBER];
+
+/*
+ * The m_allocated_channels variable stores number of currently allocated
+ * and activated watchdog channels.
+ */
+static u8_t                m_allocated_channels;
+
+/*
+ * The m_timeout variable stores watchdog timeout value in millisecond units.
+ * It is used to check whether installing watchdog provide the same timeout
+ * value in the window configuration.
+ */
+static u32_t               m_timeout;
+
+
+static void wdt_event_handler(void)
+{
+	int i;
+
+	for (i = 0; i < m_allocated_channels ; ++i) {
+		if (nrf_wdt_request_status(m_ch[i])) {
+			if (m_callbacks[i]) {
+				m_callbacks[i](DEVICE_GET(wdt_nrf), i);
+			}
+		}
+	}
+}
+
+static void wdt_nrf_isr(struct device *dev)
+{
+	ARG_UNUSED(dev);
+	/* We need to implement our own interrupt handler until nrfx one will
+	 * be fixed.
+	 * Clearing events also clears the reload register and has to be done
+	 * AFTER event handler.
+	 */
+	wdt_event_handler();
+	nrf_wdt_event_clear(NRF_WDT_EVENT_TIMEOUT);
+}
+
+static int wdt_nrf_setup(struct device *dev, u8_t options)
+{
+	nrfx_wdt_config_t config;
+	nrfx_err_t err_code;
+
+	ARG_UNUSED(dev);
+
+	if (!nrf_clock_lf_is_running()) {
+		return -ENOTSUP;
+	}
+
+	/* Activate all available options. Run in all cases. */
+	config.behaviour = NRF_WDT_BEHAVIOUR_RUN_SLEEP_HALT;
+
+	/* Deactivate running in sleep mode. */
+	if (options & WDT_OPT_PAUSE_IN_SLEEP) {
+		config.behaviour &= ~NRF_WDT_BEHAVIOUR_RUN_SLEEP;
+	}
+
+	/* Deactivate running when debugger is attached. */
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		config.behaviour &= ~NRF_WDT_BEHAVIOUR_RUN_HALT;
+	}
+
+	config.reload_value = m_timeout;
+
+	err_code = nrfx_wdt_init(&config, wdt_event_handler);
+	if (err_code == NRFX_ERROR_INVALID_STATE) {
```

Comment:
`err_code != NRFX_SUCCESS` will be safer.

Reply:
right